Provide more general factory class for FEEvaluation

kronbichler opened the pull request:
This PR adds a factory class for the `evaluate` and `integrate` calls of `FEEvaluation` and `FEFaceEvaluation`. As opposed to `SelectEvaluator` which seems to rebuild the code every time the class is used, this places the code in a separate file to force separation. It could have been achieved with that class also with some work but I did not want to repeat everything for `FEFaceEvaluation` - I have chosen a shorter approach with less code.

Right now I use a `FE_EVAL_FACTORY_DEGREE_MAX` to let the user pre-compile different degrees. We could think of making that a cmake variable to more easily change things. However, the question that is still open is the range for the quadrature formulas, which is hardcoded right now.

Whenever the code hits the pre-compiled part but with a degree not expanded, we fall back to the generic evaluate function that does not have the template (and is >2x slower).

Fixes #9794. But I must admit that we currently get a bigger library with this PR because code gets pre-compiled for all types of `VectorizedArray`.

In a follow-up PR, we should clean up the internal interfaces because it is ugly to pass around that many arguments, especially in the `gather_evaluate` and `integrate_scatter` calls.

kronbichler commented on include/deal.II/matrix_free/evaluation_template_factory.templates.h:
```diff
@@ -0,0 +1,367 @@
+
+namespace internal
+{
+  template <int degree, typename EvaluatorType, typename... Args>
+  bool
+  instantiation_helper_run(const unsigned int given_degree,
+                           const unsigned int n_q_points_1d,
+                           Args &... args)
+  {
+    if (given_degree == degree)
+      {
+        if (n_q_points_1d == degree + 1)
+          return EvaluatorType::template run<degree, degree + 1>(args...);
+        else if (n_q_points_1d == degree + 2)
+          return EvaluatorType::template run<degree, degree + 2>(args...);
+        else if (n_q_points_1d == degree)
+          return EvaluatorType::template run<degree, degree>(args...);
+        else
+          // slow path
+          return EvaluatorType::template run<-1, 0>(args...);
+      }
+    else if (degree < FE_EVAL_FACTORY_DEGREE_MAX)
+      return instantiation_helper_run<
+        (degree < FE_EVAL_FACTORY_DEGREE_MAX ? degree + 1 : degree),
+        EvaluatorType>(given_degree, n_q_points_1d, args...);
+    else
+      // slow path
+      return EvaluatorType::template run<-1, 0>(args...);
+  }
```
This is the interesting function that recursively goes into all other degrees. If we wanted more quadrature formulas we would add them here.

peterrum replied:
I think here is one quadrature formula missing which we need? Over-integration?

kronbichler replied:
I was a bit split about over-integration, but yes, I added it now. It only adds code for degrees 4 to 6 because `3*degree/2+1` gets `degree+1` for degree 1 and `degree+2` for degrees 2 and 3.

peterrum replied:
Thanks!

kronbichler commented on source/matrix_free/evaluation_template_factory.inst.in:
```diff
@@ -0,0 +1,82 @@
+
+
+for (deal_II_dimension : DIMENSIONS;
+     deal_II_scalar_vectorized : REAL_SCALARS_VECTORIZED)
```
This is the reason why it gets expensive - we need to compile it for all `VectorizedArray` types we have. We could again think of a shortcut to only use fast paths for some degrees, but I fear that we would need to be careful with describing this for the user to not give surprises when performance differs by a factor of 2.

peterrum reviewed (commented):
Looks good to me once the tests are happy. 

> In a follow-up PR, we should clean up the internal interfaces because it is ugly to pass around that many arguments, especially in the gather_evaluate and integrate_scatter calls.

Indeed, we need to do something here!

peterrum commented:
By the way, I think this is a change worth a change-log entry, isn't it?

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -7323,15 +7321,26 @@ FEEvaluation<dim,
-    this->gradients_quad,
-    this->hessians_quad,
-    this->scratch_data);
+  if (fe_degree > -1)
+    SelectEvaluator<dim, fe_degree, n_q_points_1d, VectorizedArrayType>::
+      evaluate(n_components,
+               evaluation_flags,
+               *this->data,
+               const_cast<VectorizedArrayType *>(values_array),
+               this->values_quad,
+               this->gradients_quad,
+               this->hessians_quad,
+               this->scratch_data);
+  else
+    internal::FEEvaluationFactory<dim, Number, VectorizedArrayType>::evaluate(
+      n_components,
+      evaluation_flags,
+      *this->data,
+      const_cast<VectorizedArrayType *>(values_array),
+      this->values_quad,
+      this->gradients_quad,
+      this->hessians_quad,
+      this->scratch_data);
```
I agree, but let us solve this with #10928 because that would reduce the duplication to 3 (?) arguments, at which point duplication is very likely the more readable version compared to adding a layer to collect the two directions (e.g. lambda).

peterrum replied:
I agree!

kronbichler commented:
I added two new tests for the features that are new, and switched some existing test cases to use to `-1` template path. Should be ready from my point of view.

peterrum reviewed (approved):
Looks good. I would probably merge tomorrow if there are no objections.

tjhei commented:
I don't understand the internal changes of this PR, but I feel that it would be great to have a paragraph about this feature in the matrix-free documentation somewhere (especially how FE_EVAL_FACTORY_DEGREE_MAX works).

Is there a way for the user to know if the current setup is in the "slow path"?

kronbichler commented:
> I don't understand the internal changes of this PR, but I feel that it would be great to have a paragraph about this feature in the matrix-free documentation somewhere (especially how FE_EVAL_FACTORY_DEGREE_MAX works).

Very good point. I added documentation for the parameter in FEEvaluation.

> Is there a way for the user to know if the current setup is in the "slow path"?

I guess we should provide a library function which prints the information to a stream given by the user. Let me try that in a follow-up PR - I open an issue for now to recall this and possibly extend the documentation.

peterrum commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -2148,43 +2149,50 @@ class FEEvaluationAccess<1, 1, Number, is_face, VectorizedArrayType>
+ * templates), deal.II also supports usage of this class based on the
+ * information in the element passed to the initialization. For this usage
+ * model, set the template parameter for the polynomial degree to -1 and
+ * choose an arbitrary number for the number of quadrature points. That code
```
> arbitrary number for the number of quadrature points

The value is not actually used, is it? I guess we cannot force it to be `-1` due to backwards compatibility.

kronbichler replied:
No, it is not used, but also it does not matter I believe because we replace the value here:
https://github.com/dealii/dealii/blob/1b5806b8e213f33e1773d7556a93f1efb618bdba/include/deal.II/matrix_free/evaluation_template_factory.templates.h#L55
If I remember exactly we also did not say explicitly that one needed the combination `-1,0` before, so I would think it does not matter. Or was there anything particular you had in mind that works in one case but not the other?

peterrum replied:
Thanks. No, I think it might give somehow to a user the wrong impression if `n_quadrature_points` is not forced to `-1` and he/she sets the values to `>0` and gets the *wrong* (not expected) result.

But this is not that important for this PR. I am happy with the PR. I let @tjhei merge!

tjhei commented:
> I guess we should provide a library function which prints the information to a stream given by the user. Let me try that in a follow-up PR - I open an issue for now to recall this and possibly extend the documentation.

Personally, I would want this to be on by default on anything I work on: if I use matrix-free, I care about performance. A warning message when starting is not something that would bother me. I guess we don't have this in any other place in the code, though.

peterrum commented:
@kronbichler I guess I am speaking on behalf of every `MatrixFree` user who has to deal with many `FEEvaluation` instances with different degrees and different number of quadrature points: this is a great addition (and not a minor one - although no one actually sees it)! Thank you for the lot of work of the last weeks to prepare this PR.
